jupyter-js-widgets/src/utils.ts

```diff
@@ -169,3 +171,108 @@ function escape_html(text: string): string {
     esc.textContent = text;
     return esc.innerHTML;
 };
+
+/**
+ * Takes an object 'state' and fills in buffer[i] at 'path' buffer_paths[i]
+ * where buffer_paths[i] is a list indicating where in the object buffer[i] should
+ * be placed
+ * Example: state = {a: 1, b: {}, c: [0, null]}
+ * buffers = [array1, array2]
+ * buffer_paths = [['b', 'data'], ['c', 1]]
+ * Will lead to {a: 1, b: array1, c: [0, array2]}
+ */
+export
+function put_buffers(state, buffer_paths, buffers) {
+    for (var i=0; i<buffer_paths.length; i++) {
+         // say we want to set state[x][y[z] = buffers[i]
+        var obj = state;
+        // we first get obj = state[x][y]
+        for (var j = 0; j < buffer_paths[i].length-1; j++)
+            obj = obj[buffer_paths[i][j]];
+        // and then set: obj[z] = buffers[i]
+        obj[buffer_paths[i][buffer_paths[i].length-1]] = buffers[i];
+    }
+}
+
+/**
+ * The inverse of put_buffers, return an objects with the new state where all buffers(ArrayBuffer)
+ * are removed. If a buffer is a member of an object, that object is cloned, and the key removed. If a buffer
```

Comment:
Furthermore, I think it's okay to assume that we don't have cyclic references. This translation step should really essentially be the last step before using JSON.stringify. I think these two assumptions will greatly simplify the code here. Also, if we don't have to clone, I think we can delete the underscore import at the top.